src/coreclr/System.Private.CoreLib/src/System/Diagnostics/StackTrace.CoreCLR.cs

```diff
@@ -66,30 +71,47 @@ private void CaptureStackTrace(int skipFrames, bool fNeedFileInfo, Exception? e)
             StackF.InitializeSourceInfo(fNeedFileInfo, e);
 
             _numOfFrames = StackF.GetNumberOfFrames();
+            _stackFrameHelper = StackF;
 
             if (_methodsToSkip > _numOfFrames)
                 _methodsToSkip = _numOfFrames;
 
-            if (_numOfFrames != 0)
+            _fNeedFileInfo = fNeedFileInfo;
+            _isForException = e != null;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="StackFrame"/> array for this StackTrace, possibly creating if needed.
+        /// </summary>
+        /// <returns></returns>
+        private StackFrame[] GetFramesCore()
+        {
+            if (_numOfFrames != 0 && _stackFrames == null)
             {
-                _stackFrames = new StackFrame[_numOfFrames];
+                Debug.Assert(_stackFrameHelper != null);
+
+                StackFrame[] stackFrames = new StackFrame[_numOfFrames];
 
                 for (int i = 0; i < _numOfFrames; i++)
                 {
-                    _stackFrames[i] = new StackFrame(StackF, i, fNeedFileInfo);
+                    stackFrames[i] = new StackFrame(_stackFrameHelper, i, _fNeedFileInfo);
                 }
 
                 // CalculateFramesToSkip skips all frames in the System.Diagnostics namespace,
                 // but this is not desired if building a stack trace from an exception.
-                if (e == null)
-                    _methodsToSkip += CalculateFramesToSkip(StackF, _numOfFrames);
+                if (!_isForException)
+                    _methodsToSkip += CalculateFramesToSkip(_stackFrameHelper, _numOfFrames);
 
                 _numOfFrames -= _methodsToSkip;
                 if (_numOfFrames < 0)
                 {
                     _numOfFrames = 0;
                 }
+
+                _stackFrames = stackFrames;
```

Comment:
This whole method is a data race as-implemented, and should probably have a lock around the initialization case.

Should I just add a `Lock` field and use that, or is there some other better approach?

Reply:
I do not think you need a lock as long as you avoid mutating the state and assign the result atomically at the end.